core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,220 @@
+                fieldNode.put("type", schemaType);
+                fieldNode.put("description", field.descriptionInfo().docString());
+
+                // TODO: Support for repeated fields with non-primitive types.
+                // Use "items": { ... }
+                // But unfortunately container types do not contain field infos.
+                // Maybe consider using $ref ?
+
+                // Iterate over each child field.
+                if (!field.childFieldInfos().isEmpty()) {
+                    // Set the current path to be "PREVIOUS_PATH/field.name"
+                    final String currentPath = path + "/" + field.name();
+
+                    // Mark current field as visited
+                    visited.put(fieldTypeName, currentPath);
+                    final FieldSchemaWithAdditionalProperties childProperties = generateFields(
+                            field.childFieldInfos(), visited, currentPath);
+
+                    fieldNode.set("properties", childProperties.node);
+                    fieldNode.put("additionalProperties", false);
+
+                    // Find which child properties are required.
+                    final ArrayNode required = mapper.createArrayNode();
+                    for (String requiredField : childProperties.requiredFieldNames) {
+                        required.add(requiredField);
+                    }
+
+                    fieldNode.set("required", required);
+                }
+
+                // Fill required fields for the current object.
+                if (field.requirement() == FieldRequirement.REQUIRED) {
+                    requiredFields.add(field.name());
+                }
+            }
+
+            // Set current field inside the returned object.
+            objectNode.set(field.name(), fieldNode);
+        }
+
+        return new FieldSchemaWithAdditionalProperties(objectNode, requiredFields);
+    }
+
+    /**
+     * Generate the JSON Schema for the given {@link StructInfo}.
+     * @param info struct info object.
+     * @return ObjectNode containing the JSON schema for the given struct.
+     */
+    public static ObjectNode generate(StructInfo info) {
+        final ObjectNode root = mapper.createObjectNode();
+        root.put("title", info.name())
+            .put("description", info.descriptionInfo().docString())
+            .put("additionalProperties", false)
+            .put("type", "object");
+
+        // Initialize an empty visit map, push current type to the visit map.
+        final Map<String, String> visited = new HashMap<>();
+        final String currentPath = "#";
+        visited.put(info.name(), currentPath);
+
+        final FieldSchemaWithAdditionalProperties properties = generateFields(info.fields(), visited,
+                                                                              currentPath);
+        root.set("properties", properties.node);
+
+        final ArrayNode required = mapper.createArrayNode();
```

Comment:
Required fields are required per field, e.g.

```{
  name: X,
  properties: {
    A: {...},
    B: {
       properties: {...},
       required: ["X"],
    }
  },
  required: ["A"]
}```

So passing that node is not going to help IMO. I get it, it is kind of counter-intuitive to pass required fields back. WDY think?

Reply:
I thought about creating another `arrayNode` before calling `generateFields` inside so that each property has it's own `required` field.
```java
final ArrayNode required = mapper.createArrayNode();
generateFields(methodFields, required, visited, path);

private ObjectNode generateFields(...) {
    ...
    for (FieldInfo field : fields) {
        ...
        // Iterate over each child field.
        if (!field.childFieldInfos().isEmpty()) {
            final ArrayNode required = mapper.createArrayNode();
            generateFields(field.childFieldInfos(), required, visited, currentPath);
        }
        ...
    }
    ...  
}
```
The current implementation is fine but I think we can remove the newly-added class `FieldSchemaWithAdditionalProperties` with this approach.

Reply:
I have removed field with additional properties by updating object nodes in place with generate methods.